### ISSUE TITLE
Refactor how request context is passed to rpc handlers

### ISSUE THREAD

**Issue by regiskuckaertz** [scalapb/zio-grpc#449]
Currently, upon a server call, a handler for an RPC will receive a value of type `RequestContext` in the environment. Indeed a definition:

```
service Service {
  rpc api(Input) returns (Output) {}
}
```

will result in the instantiation of a service that has the following general shape:

```
trait ZService[-context, -r] {
  def api(request: Input): ZIO[r with context, Status, Output]
}
```

which is ultimately instantiated with `context = RequestContext`. This is invoked by the underlying implementation in the server call handler:

https://github.com/scalapb/zio-grpc/blob/94987f2bd2c2a567bc4672bb8159a0ae34f8f8ef/core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala#L61-L69

where `impl` is `ZService#api`. Creating one-time environments like so has been identified as a major performance penalty: https://github.com/zio/zio/issues/7257 And it is compounded with context transformations, which manipulate the environment.

What I would suggest instead is to pass the context as a parameter:

```
trait ZService[-context, -r] {
  def api(context: context, request: Input): ZIO[r, Status, Output]
}
```

It would require to rethink the approach to transformation, but after having gone through writing a few services I have to say `r with context` is a bit awkward to work with, because `r` is usually a set of services and `context` is a value. A transform would become:

```scala
abstract class ZTransform[+rIn, +ctxIn, e, -rOut, -ctxOut] {
  def effect[a, b](f: (ctxIn, a) => ZIO[rIn, e, b]): ZIO[rOut, e, (ctxOut, a) => ZIO[rOut, e, b]]]
  def stream[a, b](f: (ctxIn, a) => ZStream[rIn, e, b]): ZIO[rOut, e, (ctxOut, a) => ZStream[rOut, e, b]]]
}
```

**Comment by thesamet** [comment]
> I know this is a significant change for the design

Can you edit the issue and add more specific details on the change you are proposing? Not sure, but I think you are proposing to see requestContext passed as an explicit parameter to all request methods in the generated code rather than passed through the environment?

**Comment by regiskuckaertz** [comment]
Yep, done!

**Comment by thesamet** [comment]
I'm generally ok with this change. Agree that working with the context through the environment can be a bit awkward, and if it is shown to have significant performance implications we can have the generated services use two parameters. As this is a breaking change, I wonder what's @ghostdogpr's current take on this (see discussion in #43)

**Comment by ghostdogpr** [comment]
In other case we have some logic that modifies the context (parses the raw context, validates it and replaces with a new one). This is achieved in a generic way by applying a transformer. If we moved the context to a single parameter, we would have to manually call the transformation for each rpc (we have hundreds 😅), right?


On the other hand if there's a significant performance improvement I can understand the motivation. I'm curious though, we're still using ZIO 1.x and this is has never showed up in any profiling. Is this with ZIO 2.x that it became a performance issue?

**Comment by regiskuckaertz** [comment]
Yes that is my understanding, you can see the ticket for details.

I think it would be just as easy to transform the parameter before being served to your implementation, it just needs a bit of tweaking to the interface.

In the synthetic benchmarks I ran, zio 1 and 2 were doing worse than other solutions, and zio 2 was the sore loser. It may have improved with the many fixes in 2.0.3 and 2.0.4, though I think there is a structural change that is needed, namely to make the core call handling completely impure.

**Comment by ghostdogpr** [comment]
Oh I missed the part about the new `ZTransform`, it would basically allowing me doing the same thing I am already doing (process the context and stick it into the environment).

**Comment by regiskuckaertz** [comment]
How about:

```scala
trait ZGeneratedService[-r, -c, s[-_, -_]] {
  this: s[r, c] =>

  def contramapContext[c1](f: c1 => c): ZGeneratedService[r, c1, s]

  def contramapContextZIO[c1](f: c1 => ZIO[r, Status, c]): ZGeneratedService[r, c1, s]
}
```

**Comment by ghostdogpr** [comment]
Sounds reasonable!

Btw is the performance much worse with zio 2 than zio 1 even if you don't transform the context? I'm about to migrate soon so quite concerned with what you said 😄 maybe we can discuss this next week at Functional Scala!

**Comment by regiskuckaertz** [comment]
I'd like to run some benchmarks with work stealing disabled, will do that this weekend. But iirc performance wise we have some work to do compared to other libraries.

**Comment by ghostdogpr** [comment]
FYI I ran the benchmarks from https://github.com/LesnyRumcajs/grpc_bench with the original version and with a version that doesn't do any `provideEnvironment`: the results are more or less the same. Most CPU time is spent in zio's FiberRuntime loop in both cases.
![Screen_Shot_2022-12-12_at_10 59 35_AM](https://user-images.githubusercontent.com/7413894/206945291-d9ed0b7d-358b-4e61-b120-6dae7b2ac6d3.png)

**Comment by regiskuckaertz** [comment]
Interesting, does that mean the recent changes in 2.0.3-4-5 have removed this problem. It remains to be seen what happens when you enable fiber roots, work stealing, etc.

**Comment by ghostdogpr** [comment]
Ah sorry, I just found out that the profiler was adding a huge overhead on each `throw` and ZIO 2 uses exceptions for control flow. After disabling exception profiling, I see the "real" CPU usage. I'll share proper details tomorrow.

**Comment by ghostdogpr** [comment]
Okay I tried stripping different things (the semaphore, getting the runtime, the provide) to measure the impact and the only one that makes a signifiant difference is actually `.race(cancelled.await)`. It goes from 38k req/s to 50 req/s with my testing conditions if I remove that line. The flamegraph profile is also quite different:

Original code:
![CPU-flame-graph](https://user-images.githubusercontent.com/7413894/207201769-d2d46112-b3ba-49a3-b285-ee471c853759.png)
Removing the race:
![CPU-flame-graph2](https://user-images.githubusercontent.com/7413894/207201767-715829bb-a8d4-4a31-b0e4-d466e60f2667.png)

**Comment by ghostdogpr** [comment]
Started a discussion on https://discord.com/channels/629491597070827530/630498701860929559/1052040972340564070

**Comment by regiskuckaertz** [comment]
wow that is quite a massive drop

**Comment by ghostdogpr** [comment]
Opened a PR to avoid using `race` and interrupt the fiber directly in case of cancellation: https://github.com/scalapb/zio-grpc/pull/457

Now even with these improvements, I think Regis's suggestion makes sense. I am only benchmarking with env = `Any`, I am guessing the overhead of the `provideSomeEnvironment` would get bigger as the environment grows so if we can avoid it, it's probably worth it.

**Comment by thesamet** [comment]
Released 0.6.0-rc1 with this change. Services come now in two flavors: one that doesn't take context, and one that takes it as a second parameter. This release is denoted `rc1` since I don't anticipate further breaking API changes. However, having said that I'm looking forward for your feedback.

**Comment by regiskuckaertz** [comment]
Oh I missed that one, great!
